include/jemalloc/internal/ehooks.h

```diff
@@ -106,18 +106,54 @@ ehooks_merge_will_fail(ehooks_t *ehooks) {
 	return ehooks_get_extent_hooks_ptr(ehooks)->merge == NULL;
 }
 
+/*
+ * Some hooks are required to return zeroed memory in certain situations.  In
+ * debug mode, we do some heuristic checks that they did what they were supposed
+ * to.
+ *
+ * This isn't really ehooks-specific (i.e. anyone can check for zeroed memory).
+ * But incorrect zero information indicates an ehook bug.
+ */
+static inline void
+ehooks_debug_zero_check(void *addr, size_t size) {
+	assert(((uintptr_t)addr & PAGE_MASK) == 0);
+	assert((size & PAGE_MASK) == 0);
+	assert(size > 0);
+	if (config_debug) {
+		/* Check the whole first page. */
+		size_t *p = (size_t *)addr;
+		for (size_t i = 0; i < PAGE / sizeof(size_t); i++) {
+			assert(p[i] == 0);
+		}
+		/* And 32 spots within. */
+		assert(PAGE >= sizeof(size_t) * 32);
+		for (size_t i = 0; i < 32; ++i) {
```

Comment:
Will do; what do you think about checking 4 (or even 2) pages instead of 32? I can think of some non-crazy bug patterns that would only zero the first page or so.

Reply:
2 or 4 pages sound fine to me.